src/Microsoft.ML.Onnx/OnnxUtils.cs

```diff
@@ -295,56 +295,40 @@ public ModelArgs(string name, TensorProto.Types.DataType dataType, List<long> di
             Contracts.CheckNonEmpty(colName, nameof(colName));
 
             TensorProto.Types.DataType dataType = TensorProto.Types.DataType.Undefined;
-            DataKind rawKind;
+            Type rawType;
             if (type is VectorType vectorType)
-                rawKind = vectorType.ItemType.RawKind;
-            else if (type is KeyType keyType)
-                rawKind = keyType.RawKind;
+                rawType = vectorType.ItemType.RawType;
+            else
+                rawType = type.RawType;
```

Comment:
>RawType [](start = 31, length = 7)

I am not sure operating over `RawType` is an appropriate alternative to `RawKind`, in this case. I might prefer that we operate over the exact `ColumnType` implementations, since that is more descriptive of what we're trying to do I think.

To give a concrete example, I see the code below is conflating `uint` as being identical, even thought the same .NET type is used for both raw `uint`s and actual key types... same .NET type, very different information being encoded. This might be a sign that the entire function is screwed up (I see @sfilipi found what appears to be a bug below with booleans, so that's well within the realm of possibility), but I'm a little concerned that this is the first thing I see.

Reply:
> I am not sure operating over RawType is an appropriate alternative to RawKind, in this case

I don't think I agree. In this case we are mapping from ML.NET's `ColumnType` to Onnx's `TensorProto.Types.DataType`. It doesn't really matter if we are using a `KeyType` or `NumberType.U4`. In either case, we need to get a `TensorProto.Types.DataType.UInt32`. Actually - what would we return for `KeyType`? We wouldn't want to hard-code it to just `TensorProto.Types.DataType.UInt32`, would we? What if it was a KeyType that was using `UInt16`?


---
In reply to: [248033212](https://github.com/dotnet/machinelearning/pull/2143#discussion_r248033212) [](ancestors = 248033212)